Add notebook with results of experiments with structured output

rlskoeser opened the pull request:
Structured output exploration related to #10 

### Changes in this PR
- New marimo notebook (and jupyter snapshot) documenting my experiments with structured outputs, using both AI Sandbox API and local ollama server

### Notes
- I recommend _not_ opening or re-running this notebook, since the results will change. I don't expect you to setup and run ollama locally, but want you to see what the output and client code looks like.  
- ollama server API has some compatibility with openai python client but I didn't test that here for structured outputs

### Evaluation
- [ ] Read through the notebook and accompanying notes and confirm that it provides enough in the way of comments, code, and output to get a sense of the structured output capabilities

laurejt commented:
If the recommendation is for the notebook to not be re-run, please change the notebook configuration so it doesn't autorun on opening.

laurejt commented on notebooks/test-structured-output.py:
```diff
@@ -0,0 +1,409 @@
+
+    system_prompt = "You are a helpful research assistant fluent in German. You help researchers identify important content in text from German scholarship."
+
+    user_prompt = f"Identify and return any passages in this page of text that quote from texts by Karl Marx:\n{page_text}"
+    return page_text, system_prompt, user_prompt
```
Consider moving the task instructions to the system prompt.

laurejt commented on notebooks/test-structured-output.py:
```diff
@@ -0,0 +1,409 @@
+def _():
+    # define some variables with sample content and prompts to use in a few different calls
+
+    page_text = """
+    Das demokrattsche Prinzip und seine Anwendung. 25 liegt das auf der Hand. Wir mögen bis an den Eingang zur Werkstatt gleich¬ sein, aber in der Werkstatt sind wir es nicht mehr. Da muß der Ingenieur anordnett und der Schlosser, Dreher &c. ausführen, da kann der Heizer nicht nach seinem Kopf verfahren und den Kessel abstellen, wenn es ihm paßt. So¬ in jedem großen Wirthschaftsunternehmen, so ab ...
+    Zinner,+etal._1896_15:01_388_Notizen,Feuilleton
+    """
```
What are the annotations / expected results for this example?

laurejt commented on notebooks/test-structured-output.py:
```diff
@@ -0,0 +1,409 @@
+            {"role": "user", "content": user_prompt},
+        ],
+        # response_format=Quote,  # not supported by our API version
+        tools=[
+            openai.pydantic_function_tool(Quote),
+        ],
```
Is this the right format given that there can be more than one quote in a passage? This seems like it'd limit the model to one quote.

laurejt commented on notebooks/test-structured-output.py:
```diff
@@ -0,0 +1,409 @@
+    from ollama import chat
+
+
+    # Supports nested models; we need to support multiple quotes on a page, so try a list of quotes
+    # using the Quote model defined earlier
+    class QuoteList(BaseModel):
+        quotes: list[Quote]
```
Can openai support nested models?

rlskoeser replied:
yes

laurejt commented on notebooks/test-structured-output.py:
```diff
@@ -0,0 +1,409 @@
+    **title:**
+    {llama_q.title}""")
+
+    mo.md("\n\n".join(llama_quotes_md))
```
Quotes are not properly separated by linebreaks, would a third newline fix this?

laurejt commented on notebooks/test-structured-output.py:
```diff
@@ -0,0 +1,409 @@
+    {llama_q.title}""")
+
+    mo.md("\n\n".join(llama_quotes_md))
+    return llama_q, llama_quotes_md
```
How good are these results?

laurejt commented on notebooks/test-structured-output.py:
```diff
@@ -0,0 +1,409 @@
+    {mixtral_q.title}""")
+
+    mo.md("\n\n".join(mixtral_quotes_md))
+    return mixtral_q, mixtral_quotes_md
```
How good are these results?

laurejt commented on notebooks/test-structured-output.py:
```diff
@@ -0,0 +1,409 @@
+    **title:**
+    {mixtral_q.title}""")
+
+    mo.md("\n\n".join(mixtral_quotes_md))
```
Quotes are not properly separated by linebreaks, would a third newline fix this?

laurejt commented on notebooks/test-structured-output.py:
```diff
@@ -0,0 +1,409 @@
+def _(mo):
+    mo.md(
+        r"""
+        ## CSV output?
+
+        CSV output might be a simpler way to get structured output, but may be less reliable.
+        """
```
What models support this output format?

laurejt reviewed (changes requested):
Although it seems like this notebook is meant more for notes rather than running, I think it's important to provide the option for running it. So, the README should include instructions for any additional setup that's needed for this notebook and any (optional) dependencies should be added to the pyproject.toml.

rlskoeser commented:
@laurejt I've updated the notebook to address some of your comments.
- added notes about the expected quotations to be found based on annotation data
- added comments about the quality of the results
- update to use nested quote lists everywhere
- improved in-notebook instructions for ollama setup
- tested the openai client with ollama for structured response (doesn't fully work)
- tested gpt-4o and 4o-mini with csv; the csv results seem to be worse than the structured data (although I did also test with a smaller passage ...)
- saved an html copy for viewing

I started adding to pyproject.toml but it didn't seem appropriate - it's only one python package and several manual steps. Unless we decide to use any of it elsewhere, let's keep it in this notebook.

I apologize about the jupyter formatting, I hadn't been looking at it closely. I think the html is more readable in some ways - although probably not easily viewable on github.

laurejt commented on notebooks/test-structured-output.py:
```diff
@@ -1,7 +1,11 @@
+app = marimo.App(
+    width="medium",
+    app_title="Test Structured Outputs",
+    css_file="custom.css",
```
Is this an actual file? If so, we should probably add it to the repo as well.

laurejt commented on notebooks/test-structured-output.py:
```diff
@@ -162,24 +187,43 @@ def _(gpt4o_completion, mo):
+
+        Correctly returns the full text of the expected quote and only that quote.
+
+        It looks like it is using the article title (?) at the top of the page text as the title.  The tag in the annotation data is "Manifest der Kommunistischen Partei"
+        """
```
Looks like the title is set to `None` in this most recent run. Perhaps just mention its variable behavior?

laurejt commented on notebooks/test-structured-output.py:
```diff
@@ -211,16 +255,33 @@ def _(gpt4omini_completion, mo):
+
+        On a previous run it returned incorrect quote: a different set of text surrounded by „“.  This time it returned the correct quote.
+
+        Like GPT4o, it returns the text at the beginning of the page as a title.
```
Again, it returns `None` for the title.

laurejt commented on notebooks/test-structured-output.py:
```diff
@@ -364,22 +431,91 @@ def _(mixtral_quotes, mo):
+
+
+@app.cell
+def _(ollama_completion):
+    print(ollama_completion.model_dump_json(indent=2))
+    return
```
For some reason in the ipynb this looks like an error traceback, but it looks fine in the html version

laurejt commented on notebooks/test-structured-output.py:
```diff
@@ -0,0 +1,617 @@
+
+@app.cell
+def _(mo):
+    mo.md(r"""### AI sandbox - gpt-4o-min""")
```
```suggestion
    mo.md(r"""### AI sandbox - gpt-4o-mini""")
```
